syntax = "proto3";

package flexim_connect;

// EchoRequest is the request for echo.
message CreateBagRequest {
  string name = 1;
}

// EchoResponse is the response for echo.
message CreateBagResponse {
  uint64 id = 1;
  string name = 2;
}

enum DataType {
  Image = 0;
  Tensor2D = 1;
  DataFrame = 2;
  Object = 3;
}

message AppendDataRequest {
  oneof data {
    DataMeta meta = 1;
    bytes data_bytes = 2;
  }

  message DataMeta {
    uint64 bag_id = 1;
    string name = 2;
    DataType data_type = 4;

    enum SpecialColumn {
      Rectangle = 0;
      Segment = 1;
      Color = 2;
      Vector = 3;
      Point = 4;
    }

    map<string, SpecialColumn> special_columns = 5;
  }
}


message AppendDataResponse {
  uint64 bag_id = 1;
  uint64 data_id = 2;
  uint64 data_size = 3;
}


message ListBagsRequest {
}

message ListBagsResponse {
  repeated BagMeta bag_metas = 1;

  message BagMeta {
    uint64 id = 1;
    string name = 2;
    uint64 number_of_data = 3;
    string created_at = 4;
  }
}

service FleximConnect {
  rpc CreateBag(CreateBagRequest) returns (CreateBagResponse) {}
  rpc AppendData(stream AppendDataRequest) returns (AppendDataResponse) {}
  rpc ListBags(ListBagsRequest) returns (ListBagsResponse) {}
}